crates/pop-fork/src/rpc.rs

```diff
@@ -0,0 +1,326 @@
+
+		Ok(Self { legacy, endpoint: endpoint.clone() })
+	}
+
+	/// Get the endpoint URL this client is connected to.
+	pub fn endpoint(&self) -> &Url {
+		&self.endpoint
+	}
+
+	/// Get the latest finalized block hash.
+	///
+	/// This is typically the starting point for forking - we fork from the latest
+	/// finalized state to ensure consistency.
+	pub async fn finalized_head(&self) -> Result<H256, RpcClientError> {
+		self.legacy
+			.chain_get_finalized_head()
+			.await
+			.map_err(|e| RpcClientError::RequestFailed(e.to_string()))
+	}
+
+	/// Get block header by hash.
+	///
+	/// Returns the header for the specified block, which contains the parent hash,
+	/// state root, extrinsics root, and digest.
+	pub async fn header(
+		&self,
+		hash: H256,
+	) -> Result<<PolkadotConfig as subxt::Config>::Header, RpcClientError> {
+		self.legacy
+			.chain_get_header(Some(hash))
+			.await
+			.map_err(|e| RpcClientError::RequestFailed(e.to_string()))?
+			.ok_or_else(|| RpcClientError::InvalidResponse(format!("No header found for {hash:?}")))
+	}
+
+	/// Get a single storage value at a specific block.
+	///
+	/// # Arguments
+	/// * `key` - The storage key (raw bytes)
+	/// * `at` - The block hash to query state at
+	///
+	/// # Returns
+	/// * `Ok(Some(value))` - Storage exists with value
+	/// * `Ok(None)` - Storage key doesn't exist (empty)
+	/// * `Err(_)` - RPC error
+	pub async fn storage(&self, key: &[u8], at: H256) -> Result<Option<Vec<u8>>, RpcClientError> {
+		self.legacy
+			.state_get_storage(key, Some(at))
+			.await
+			.map_err(|e| RpcClientError::RequestFailed(e.to_string()))
+	}
+
+	/// Get multiple storage values in a single batch request.
+	///
+	/// More efficient than multiple individual `storage()` calls when fetching
+	/// many keys at once.
+	///
+	/// # Arguments
+	/// * `keys` - Slice of storage keys to fetch
+	/// * `at` - The block hash to query state at
+	///
+	/// # Returns
+	/// A vector of optional values, in the same order as the input keys.
+	pub async fn storage_batch(
+		&self,
+		keys: &[Vec<u8>],
+		at: H256,
+	) -> Result<Vec<Option<Vec<u8>>>, RpcClientError> {
+		if keys.is_empty() {
+			return Ok(vec![]);
+		}
+
+		// Use state_queryStorageAt for batch fetching
+		let keys_refs: Vec<&[u8]> = keys.iter().map(|k| k.as_slice()).collect();
+
+		let result = self
+			.legacy
+			.state_query_storage_at(keys_refs, Some(at))
+			.await
+			.map_err(|e| RpcClientError::RequestFailed(e.to_string()))?;
+
+		// Build a map of key -> value from the response
+		let changes: std::collections::HashMap<Vec<u8>, Option<Vec<u8>>> = result
+			.into_iter()
+			.flat_map(|change_set| {
+				change_set.changes.into_iter().map(|(k, v)| {
+					let key_bytes = k.0.to_vec();
+					let value_bytes = v.map(|v| v.0.to_vec());
+					(key_bytes, value_bytes)
+				})
+			})
+			.collect();
+
+		// Return values in the same order as input keys
+		let values = keys.iter().map(|key| changes.get(key).cloned().flatten()).collect();
```

Comment:
values can be arbitrarily big, these clones may be more expensive than what we can imagine. I'd just remove the ownership from the map and use always the same memory blob.

```suggestion
		let values = keys.iter().map(|key| changes.remove(key)).collect();
```